lib/internal/modules/sync_hooks.js

```diff
@@ -0,0 +1,338 @@
+'use strict';
+
+const {
+  ArrayPrototypeFindIndex,
+  ArrayPrototypePush,
+  ArrayPrototypeSplice,
+  ObjectFreeze,
+  StringPrototypeStartsWith,
+  Symbol,
+} = primordials;
+const {
+  isAnyArrayBuffer,
+  isArrayBufferView,
+} = require('internal/util/types');
+
+const { BuiltinModule } = require('internal/bootstrap/realm');
+const {
+  ERR_INVALID_RETURN_PROPERTY_VALUE,
+} = require('internal/errors').codes;
+const { validateFunction } = require('internal/validators');
+const { isAbsolute } = require('path');
+const { pathToFileURL, fileURLToPath } = require('internal/url');
+
+let debug = require('internal/util/debuglog').debuglog('module_hooks', (fn) => {
+  debug = fn;
+});
+
+/** @typedef {import('internal/modules/cjs/loader.js').Module} Module */
+
+// Use arrays for better insertion and iteration performance, we don't care
+// about deletion performance as much.
+const resolveHooks = [];
+const loadHooks = [];
+const hookId = Symbol('kModuleHooksIdKey');
+let nextHookId = 0;
+
+class ModuleHooks {
+  constructor(resolve, load) {
+    this[hookId] = Symbol(`module-hook-${nextHookId++}`);
+    // Always initialize all hooks, if it's unspecified it'll be an owned undefined.
+    this.resolve = resolve;
+    this.load = load;
+
+    if (resolve) {
+      ArrayPrototypePush(resolveHooks, this);
+    }
+    if (load) {
+      ArrayPrototypePush(loadHooks, this);
+    }
+
+    ObjectFreeze(this);
+  }
+  // TODO(joyeecheung): we may want methods that allow disabling/enabling temporarily
+  // which just sets the item in the array to undefined temporarily.
+  // TODO(joyeecheung): this can be the [Symbol.dispose] implementation to pair with
+  // `using` when the explicit resource management proposal is shipped by V8.
+  deregister() {
```

Comment:
JSDoc isn’t just about types, it’s also to explain the purpose of the function. Deregister what? A particular hook? All hooks? It’s not obvious from the method signature, since the method takes no input.

In general though we decided a while back to require JSDoc comments for all functions and methods in the modules codebase, enforced via lint rule, so I’m also making a suggestion that I would assume should help the lint check pass.

Reply:
I think it being an instance method is already self-explanatory and I find redundant comments making the code harder to read (one of the problems I find with the module loader code is that a lot of comments are only there for comment's sake without providing any additional information that can't already be understood from naming). But sure I can add a sentence explaining it deregister the instance.